osu.Game/Screens/Select/BeatmapDetails.cs

```diff
@@ -0,0 +1,513 @@
+using osu.Game.Graphics.UserInterface;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+
+namespace osu.Game.Screens.Select
+{
+    public class BeatmapDetails : Container
+    {
+        private readonly OsuSpriteText description;
+        private readonly OsuSpriteText source;
+        private readonly FillFlowContainer<OsuSpriteText> tags;
+
+        private readonly DifficultyRow circleSize;
+        private readonly DifficultyRow drainRate;
+        private readonly DifficultyRow overallDifficulty;
+        private readonly DifficultyRow approachRate;
+        private readonly DifficultyRow stars;
+
+        private readonly Bar ratingsBar;
+        private readonly OsuSpriteText negativeRatings;
+        private readonly OsuSpriteText positiveRatings;
+        private readonly BarGraph ratingsGraph;
+
+        private readonly BarGraph retryGraph;
+        private readonly BarGraph failGraph;
+
+        private BeatmapInfo beatmap;
+
+        public BeatmapInfo Beatmap
+        {
+            get
+            {
+                return beatmap;
+            }
+
+            set
+            {
+                if (beatmap == value) return;
+                beatmap = value;
+                description.Text = beatmap.Version;
+                source.Text = beatmap.Metadata.Source;
+                tags.Children = beatmap.Metadata.Tags?.Split(' ').Select(text => new OsuSpriteText
+                {
+                    Text = text,
+                    Font = "Exo2.0-Regular",
+                });
+
+                circleSize.Value = beatmap.Difficulty.CircleSize;
+                drainRate.Value = beatmap.Difficulty.DrainRate;
+                overallDifficulty.Value = beatmap.Difficulty.OverallDifficulty;
+                approachRate.Value = beatmap.Difficulty.ApproachRate;
+                stars.Value = (float)beatmap.StarDifficulty;
+            }
+        }
+
+        private List<int> ratings;
+
+        public IEnumerable<int> Ratings
+        {
+            get
+            {
+                return ratings;
+            }
+            set
+            {
+                ratings = value.ToList();
+                negativeRatings.Text = ratings.GetRange(0, 5).Sum().ToString();
+                positiveRatings.Text = ratings.GetRange(5, 5).Sum().ToString();
+                ratingsBar.Length = (float)ratings.GetRange(0, 5).Sum() / ratings.Sum();
+
+                ratingsGraph.Values = ratings.Select(rating => (float)rating);
+            }
+        }
+
+        private List<int> retries = Enumerable.Repeat(0, 100).ToList();
+
+        public IEnumerable<int> Retries
+        {
+            get
+            {
+                return retries;
+            }
+            set
+            {
+                retries = value.ToList();
+                calcRetryAndFailGraph();
+            }
+        }
+
+        private List<int> fails = Enumerable.Repeat(0, 100).ToList();

```

Comment:
Am not sure about these defaults. Maybe just make them null and then nullcheck appropriately. This should fit in well with the other comment I made, about how sections that have no data should be hidden.